File: apps/api/src/middleware/rate-limit.ts
import { createMiddleware } from "hono/factory";
import type { AppEnv } from "../context";
import { sha256Hex } from "../db/scoped";

/**
 * The client's address, as far as it can be trusted.
 *
 * On Workers this was simply cf-connecting-ip, which Cloudflare sets and a
 * caller cannot forge. There is no such header off Cloudflare, and
 * X-Forwarded-For is caller-supplied: trusting it blindly would let anyone
 * mint a fresh rate-limit bucket per request and walk straight through the
 * brake.
 *
 * So the header is only consulted when the operator has declared how many
 * proxies sit in front (TRUSTED_PROXY_HOPS), and the address is counted from
 * the RIGHT — the last entry a trusted proxy actually observed. Anything a
 * client prepends sits further left and is ignored.
 *
 * With 0 hops (the default) the header is not read at all.
 */
export function clientIp(
  forwardedFor: string | null,
  socketAddress: string | null,
  trustedHops: number,
): string {
  if (trustedHops <= 0) return socketAddress ?? "unknown";
  const chain = (forwardedFor ?? "")
    .split(",")
    .map((part) => part.trim())
    .filter(Boolean);
  if (chain.length === 0) return socketAddress ?? "unknown";
  // The rightmost entry was appended by the nearest proxy, so hop N back from
  // the end is the address the outermost trusted proxy saw.
  const index = chain.length - trustedHops;
  return chain[Math.max(0, index)] ?? socketAddress ?? "unknown";
}

// Fixed-window counter. scope "ip" buckets per client; "global" is one shared
// bucket across all clients (defeats distributed guessing at the cost of
// shared-fate 429s — use generous limits).
export function rateLimit(opts: {
  name: string;
  limit: number;
  windowSeconds: number;
  scope?: "ip" | "global";
}) {
  return createMiddleware<AppEnv>(async (c, next) => {
    const ip =
      opts.scope === "global"
        ? "global"
        : clientIp(
            c.req.header("x-forwarded-for") ?? null,
            // The peer address, which only the Bun server handle knows.
            // Absent (in tests, or before the server is listening) callers
            // share the "unknown" bucket — safe, just coarse.
            c.var.deps.peerAddress(c.req.raw),
            c.var.deps.trustedProxyHops,
          );
    // Hashed, never the address itself. KV forced this — it was globally
    // replicated with no jurisdiction option — and the counters now live in
    // the same EU database as everything else, but there is still no reason
    // to record addresses, and a hash buckets each client identically.
    const bucket =
      ip === "global" ? "global" : (await sha256Hex(ip)).slice(0, 32);
    const window = Math.floor(Date.now() / 1000 / opts.windowSeconds);
    const key = `rl:${opts.name}:${bucket}:${window}`;
    // One atomic increment. The KV version read, compared and wrote back,
    // which was racy by nature ("a brake, not an invariant"); with several
    // replicas sharing one database that race would have got materially
    // worse, so the counter is now exact.
    const count = await c.var.rateLimit.hit(key, opts.windowSeconds);
    if (count > opts.limit) {
      return c.json(
        { error: "Too many attempts, try again later", code: "RATE_LIMITED" },
        429,
      );
    }
    await next();
  });
}

File: apps/server/internal/ratelimit/ratelimit.go
// Package ratelimit holds Pjokk's rate-limit counter store (the `rate_limit`
// table, CLAUDE.md) and the client-address resolution the buckets are keyed
// on.
//
// See docs/superpowers/plans/2026-08-31-go-migration-reference.md §A5 items
// 7-8 and §A6, where the port is named RateLimitStore; the package name here
// already carries "ratelimit", so the exported type is just Store.
//
// The counters used to live in Cloudflare KV, which was eventually consistent:
// the old limiter read a value, compared it and wrote it back, a race its own
// comment accepted as "a brake, not an invariant". Postgres removes the
// compromise — one statement increments and returns the new value atomically,
// so the limit holds even when several replicas serve the same caller at once.
package ratelimit

import (
	"context"
	"fmt"
	"net"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgtype"

	"github.com/refsdal/pjokk/server/internal/db/gen"
)

// Store is the rate-limit counter port.
type Store interface {
	// Hit increments the counter for key within a sliding window of
	// windowSeconds and returns the count after incrementing.
	Hit(ctx context.Context, key string, windowSeconds int) (int, error)
	// Sweep deletes expired counters as of now, returning how many were
	// removed.
	Sweep(ctx context.Context, now time.Time) (int, error)
}

// Postgres is the shipped Store: one row per (limiter, bucket, window) in the
// `rate_limit` table.
type Postgres struct {
	q *gen.Queries
}

var _ Store = (*Postgres)(nil)

// NewPostgres builds the store over an existing querier, so it shares the one
// pool the composition root opened.
func NewPostgres(q *gen.Queries) *Postgres { return &Postgres{q: q} }

// Hit increments the counter for key and returns the new count.
//
// The row's expiry is kept generous — at least a minute, otherwise twice the
// window — so a clock skew between replicas cannot resurrect a bucket that
// should already have expired. Nothing reads expires_at to decide whether a
// bucket counts: the window number is part of the key, and the expiry exists
// only so Sweep has something to prune by.
func (p *Postgres) Hit(ctx context.Context, key string, windowSeconds int) (int, error) {
	ttl := windowSeconds * 2
	if ttl < 60 {
		ttl = 60
	}
	count, err := p.q.HitRateLimit(ctx, gen.HitRateLimitParams{
		Key:       key,
		ExpiresAt: pgtype.Timestamptz{Time: time.Now().Add(time.Duration(ttl) * time.Second), Valid: true},
	})
	if err != nil {
		return 0, fmt.Errorf("ratelimit: hit %q: %w", key, err)
	}
	return int(count), nil
}

// Sweep deletes every counter that expired before now.
func (p *Postgres) Sweep(ctx context.Context, now time.Time) (int, error) {
	removed, err := p.q.SweepRateLimit(ctx, pgtype.Timestamptz{Time: now, Valid: true})
	if err != nil {
		return 0, fmt.Errorf("ratelimit: sweep: %w", err)
	}
	return int(removed), nil
}

// ClientIP is the client's address, as far as it can be trusted (REF §A5
// item 7).
//
// On Workers this was simply cf-connecting-ip, which Cloudflare set and a
// caller could not forge. There is no such header off Cloudflare, and
// X-Forwarded-For is caller-supplied: trusting it blindly would let anyone
// mint a fresh rate-limit bucket per request and walk straight through the
// brake.
//
// So the header is only consulted when the operator has declared how many
// proxies sit in front (TRUSTED_PROXY_HOPS), and the address is counted from
// the RIGHT — the last entry a trusted proxy actually observed. Anything a
// client prepends sits further left and is ignored. With 0 hops (the default)
// the header is not read at all.
//
// Divergence from the TypeScript original: there, socketAddress came from
// Bun's `server.requestIP().address`, which is a bare IP. In Go the socket
// address arrives as net/http's r.RemoteAddr, which carries a port — and a
// port would put every single connection in its own bucket, quietly disabling
// the limiter. The port is therefore stripped here.
func ClientIP(forwardedFor string, socketAddress string, trustedHops int) string {
	socket := stripPort(socketAddress)

	if trustedHops <= 0 {
		return orUnknown(socket)
	}

	chain := make([]string, 0, 4)
	for _, part := range strings.Split(forwardedFor, ",") {
		if trimmed := strings.TrimSpace(part); trimmed != "" {
			chain = append(chain, trimmed)
		}
	}
	if len(chain) == 0 {
		return orUnknown(socket)
	}

	// The rightmost entry was appended by the nearest proxy, so hop N back
	// from the end is the address the outermost trusted proxy saw.
	index := len(chain) - trustedHops
	if index < 0 {
		index = 0
	}
	return orUnknown(chain[index])
}

// stripPort removes a trailing port from an address, leaving IPv6 literals
// unbracketed. An address without a port is returned unchanged.
func stripPort(address string) string {
	if address == "" {
		return ""
	}
	if host, _, err := net.SplitHostPort(address); err == nil {
		return host
	}
	return address
}

func orUnknown(address string) string {
	if address == "" {
		return "unknown"
	}
	return address
}
